### PR TITLE
[SPARK-48572][SQL] Fix DateSub, DateAdd, WindowTime, TimeWindow and SessionWindow expressions

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
Fix for listed expressions.

### Why are the changes needed?
These expressions are found to be faulty when working with collations.

### Does this PR introduce _any_ user-facing change?
Yes, it fixes expressions that are not working with collations.

### How was this patch tested?
Tests added for expressions to reproduce errors.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:1]: note: I wanted to eventually cover this in https://github.com/apache/spark/pull/46659, but I'll close that now

**Comment by uros-db** [comment]
btw, should we add some tests for these changes?

**Comment by mihailom-db** [comment]
Will add them a bit later and will remove WIP when it is done.

**Review by uros-db — APPROVED** [review]
lgtm

**Comment by mihailom-db** [comment]
@cloud-fan could you review this?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**